subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+			(struct audio_module_handle_private *)handle, configuration);
+		if (ret) {
+			LOG_DBG("Set configuration for module %s send failed, ret %d", handle->name,
+				ret);
+			return ret;
+		}
+	}
+
+	switch (handle->description->type) {
+	case AMOD_TYPE_INPUT:
+		thread_entry = (k_thread_entry_t)module_thread_input;
+		break;
+
+	case AMOD_TYPE_OUTPUT:
+		thread_entry = (k_thread_entry_t)module_thread_output;
+		break;
+
+	case AMOD_TYPE_PROCESS:
+		thread_entry = (k_thread_entry_t)module_thread_in_out;
+		break;
+
+	default:
+		LOG_DBG("Invalid module type %d for module %s", handle->description->type,
+			handle->name);
+		return -EINVAL;
+	}
+
+	sys_slist_init(&handle->hdl_dest_list);
+	k_mutex_init(&handle->dest_mutex);
+
+	handle->thread_id = k_thread_create(
+		&handle->thread_data, handle->thread.stack, handle->thread.stack_size,
+		(k_thread_entry_t)module_thread_in_out, (void *)handle, NULL, NULL,
+		K_PRIO_PREEMPT(handle->thread.priority), 0, K_FOREVER);
+
+	ret = k_thread_name_set(handle->thread_id, &handle->name[0]);
+	if (ret) {
+		LOG_DBG("Failed to start thread for module %s thread, ret %d", handle->name, ret);
+
+		/* Clean up the handle */
+		memset(handle, 0, sizeof(struct audio_module_handle));
+		return ret;
+	}
+
+	handle->state = AMOD_STATE_CONFIGURED;
+
+	k_thread_start(handle->thread_id);
+
+	LOG_DBG("Start thread");
+
+	return 0;
+}
+
+/**
+ * @brief  Function to close an open module.
+ */
+int audio_module_close(struct audio_module_handle *handle)
+{
+	int ret;
+
+	if (handle == NULL) {
+		LOG_DBG("Module handle is NULL");
+		return -EINVAL;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED || handle->state == AMOD_STATE_RUNNING) {
+		LOG_DBG("Module %s in an invalid state, %d, for close", handle->name,
+			handle->state);
+		return -ENOTSUP;
+	}
+
+	if (handle->description->functions->close != NULL) {
+		ret = handle->description->functions->close(
+			(struct audio_module_handle_private *)handle);
+		if (ret) {
+			LOG_DBG("Failed close call to module %s, returned %d", handle->name, ret);
+			return ret;
+		}
+	}
+
+	if (handle->thread.msg_rx != NULL) {
+		data_fifo_empty(handle->thread.msg_rx);
+	}
+
+	if (handle->thread.msg_tx != NULL) {
+		data_fifo_empty(handle->thread.msg_tx);
+	}
+
+	/*
+	 * How to return all the data to the slab items?
+	 * Test the semaphore and wait for it to be zero.
+	 */
+
+	k_thread_abort(handle->thread_id);
+
+	LOG_DBG("Closed module %s", handle->name);
+
+	/* Clean up the handle */
+	memset(handle, 0, sizeof(struct audio_module_handle));
+
+	return 0;
+};
+
+/**
+ * @brief  Function to reconfigure a module.
+ */
+int audio_module_reconfigure(struct audio_module_handle *handle,
+			     struct audio_module_configuration *configuration)
+{
+	int ret;
+
+	if (handle == NULL || configuration == NULL) {
+		LOG_DBG("Module input parameter error");
+		return -EINVAL;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED || handle->state == AMOD_STATE_RUNNING) {
```

Comment:
Can this be set for all states besides UNDEFINED and RUNNING?

Reply:
Yes as the module is in the stable states of configured or stopped.